File: packages/client/src/pda.ts
import type { Address, ProgramDerivedAddress } from "@solana/kit";
import { getAddressEncoder, getProgramDerivedAddress } from "@solana/kit";

import { TUNA_PROGRAM_ADDRESS } from "./generated";

export async function getTunaConfigAddress(): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["tuna_config"],
  });
}

export async function getMarketAddress(pool: Address): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["market", getAddressEncoder().encode(pool)],
  });
}

export async function getLendingVaultAddress(mint: Address): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["vault", getAddressEncoder().encode(mint)],
  });
}

export async function getLendingPositionAddress(wallet: Address, mint: Address): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["lending_position", getAddressEncoder().encode(wallet), getAddressEncoder().encode(mint)],
  });
}

export async function getTunaPositionAddress(position_mint: Address): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["tuna_position", getAddressEncoder().encode(position_mint)],
  });
}

File: packages/client-rs/src/pda.rs
use solana_program::pubkey::Pubkey;
use crate::TUNA_ID;

pub fn get_tuna_position_address(position_mint: &Pubkey) -> (Pubkey, u8) {
    let seeds = &[b"tuna_position", position_mint.as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

pub fn get_market_address(whirlpool: &Pubkey) -> (Pubkey, u8) {
    let seeds = &[b"market", whirlpool.as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

pub fn get_tuna_config_address() -> (Pubkey, u8) {
    let seeds = &[b"tuna_config".as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

pub fn get_vault_address(token_mint: &Pubkey) -> (Pubkey, u8) {
    let seeds = &[b"vault", token_mint.as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}